src/julia-syntax.scm

```diff
@@ -3382,12 +3388,34 @@ f(x) = yt(x)
       `(call (core getfield) ,fname ,(get opaq var))
       `(call (core getfield) ,fname (inert ,var))))
 
+(define (convert-global-assignment var rhs0 globals)
+  (let* ((rhs1 (if (or (simple-atom? rhs0)
+                       (equal? rhs0 '(the_exception)))
+                   rhs0
+                   (make-ssavalue)))
+         (ref   (binding-to-globalref var))
+         (ty   `(call (core get_binding_type) ,(cadr ref) (inert ,(caddr ref))))
+         (rhs  (if (get globals ref #t) ;; no type declaration for constants
+                   (let ((ex (convert-for-type-decl rhs1 ty)))
+                     (if (has? globals ref)
+                         ex
+                         `(toplevel-butfirst
+                            ,ex
+                            (call (core set_binding_type!) ,(cadr ref) (inert ,(caddr ref))))))
```

Comment:
Could we call this in `resolve_globals` instead (where other `global x` declarations in functions are handled), so the front end doesn't need to handle it?

Reply:
Hmm, yes, maybe. I don't know if we can distinguish writes from only reads from globals there though.

Reply:
`global x` is the same as a write, for syntactic purposes

Reply:
Ok, I tried that, but the issue are toplevel expressions like `x = 1`. Those are writes to a global variable, but this is difficult to recognize in the interpreter before `get_binding_type` is evaluated, since there is no `global` expression.

Reply:
Maybe instead of calling get_binding_type, the front end could call get_binding_type_for_write, which sets the type to Any if there isn't one yet.

Reply:
Yeah, that makes sense. I think we can just change he semantics of `get_binding_type` that way, since it should only be called through the frontend anyways.

Reply:
Ok, what do you think of [a334a9b](https://github.com/JuliaLang/julia/pull/43671/commits/a334a9b06bcfe47e59d0d50dc82499d2f41f6461)?